File: apps/server/AliasVault.Client/Main/Models/PasswordSettings.cs
//-----------------------------------------------------------------------
// <copyright file="PasswordSettings.cs" company="aliasvault">
// Copyright (c) aliasvault. All rights reserved.
// Licensed under the AGPLv3 license. See LICENSE.md file in the project root for full license information.
// </copyright>
//-----------------------------------------------------------------------

namespace AliasVault.Client.Main.Models;

using System.Text.Json.Serialization;
using AliasClientDb.Models;

/// <summary>
/// Settings for password generation.
/// </summary>
public class PasswordSettings
{
    /// <summary>
    /// Gets or sets the length of the password.
    /// </summary>
    [JsonPropertyName("Length")]
    public int Length { get; set; } = PasswordGeneratorDefaults.DefaultPasswordLength;

    /// <summary>
    /// Gets or sets a value indicating whether to use lowercase letters.
    /// </summary>
    [JsonPropertyName("UseLowercase")]
    public bool UseLowercase { get; set; } = true;

    /// <summary>
    /// Gets or sets a value indicating whether to use uppercase letters.
    /// </summary>
    [JsonPropertyName("UseUppercase")]
    public bool UseUppercase { get; set; } = true;

    /// <summary>
    /// Gets or sets a value indicating whether to use numbers.
    /// </summary>
    [JsonPropertyName("UseNumbers")]
    public bool UseNumbers { get; set; } = true;

    /// <summary>
    /// Gets or sets a value indicating whether to use special characters.
    /// </summary>
    [JsonPropertyName("UseSpecialChars")]
    public bool UseSpecialChars { get; set; } = true;

    /// <summary>
    /// Gets or sets a value indicating whether to use non-ambiguous characters.
    /// </summary>
    [JsonPropertyName("UseNonAmbiguousChars")]
    public bool UseNonAmbiguousChars { get; set; } = false;

    /// <summary>
    /// Gets or sets which generator to use ("basic" or "diceware"). Defaults to "basic".
    /// </summary>
    [JsonPropertyName("Type")]
    public string Type { get; set; } = "basic";

    /// <summary>
    /// Gets or sets the number of words in a Diceware passphrase.
    /// </summary>
    [JsonPropertyName("WordCount")]
    public int WordCount { get; set; } = PasswordGeneratorDefaults.DefaultWordCount;

    /// <summary>
    /// Gets or sets the Diceware wordlist language. Empty means "auto": the most appropriate available
    /// wordlist is resolved from the app language at runtime.
    /// </summary>
    [JsonPropertyName("Language")]
    public string Language { get; set; } = string.Empty;

    /// <summary>
    /// Gets or sets the Diceware capitalization ("None", "TitleCase", "Uppercase", "Lowercase" or "Random").
    /// </summary>
    [JsonPropertyName("Capitalization")]
    public string Capitalization { get; set; } = "Lowercase";

    /// <summary>
    /// Gets or sets the separator between Diceware words ("None", "Dash", "Space", "Underscore" or "Dot").
    /// </summary>
    [JsonPropertyName("Separator")]
    public string Separator { get; set; } = "Dash";

    /// <summary>
    /// Gets or sets the optional random salt character for a Diceware passphrase ("None", "Prefix", "Sprinkle" or "Suffix").
    /// </summary>
    [JsonPropertyName("Salt")]
    public string Salt { get; set; } = "None";
}

File: core/models/src/vault/PasswordSettings.ts
/**
 * Which generator a password setting uses.
 */
export type PasswordGeneratorType = 'basic' | 'diceware';

/**
 * Capitalization applied to each Diceware word.
 */
export type DicewareCapitalization = 'None' | 'TitleCase' | 'Uppercase' | 'Lowercase' | 'Random';

/**
 * Separator placed between Diceware words.
 */
export type DicewareSeparator = 'None' | 'Dash' | 'Space' | 'Underscore' | 'Dot';

/**
 * Optional random salt character added to a Diceware passphrase.
 */
export type DicewareSalt = 'None' | 'Prefix' | 'Sprinkle' | 'Suffix';

/**
 * Settings for password generation stored in SQLite database settings table as string.
 *
 * The Diceware fields are optional so that older stored blobs (which only contain the
 * basic-password fields) remain valid; defaults are applied when reading the settings.
 */
export type PasswordSettings = {
  /**
   * The length of the password (basic generator).
   */
  Length: number;

  /**
   * Whether to use lowercase letters (basic generator).
   */
  UseLowercase: boolean;

  /**
   * Whether to use uppercase letters (basic generator).
   */
  UseUppercase: boolean;

  /**
   * Whether to use numbers (basic generator).
   */
  UseNumbers: boolean;

  /**
   * Whether to use special characters (basic generator).
   */
  UseSpecialChars: boolean;

  /**
   * Whether to use non-ambiguous characters (basic generator).
   */
  UseNonAmbiguousChars: boolean;

  /**
   * Which generator to use. Defaults to 'basic' when absent.
   */
  Type?: PasswordGeneratorType;

  /**
   * Number of words in the passphrase (diceware generator).
   */
  WordCount?: number;

  /**
   * Wordlist language code (diceware generator).
   */
  Language?: string;

  /**
   * Capitalization applied to each word (diceware generator). Defaults to 'Lowercase'.
   */
  Capitalization?: DicewareCapitalization;

  /**
   * Separator between words (diceware generator). Defaults to 'Dash'.
   */
  Separator?: DicewareSeparator;

  /**
   * Optional random salt character (diceware generator). Defaults to 'None'.
   */
  Salt?: DicewareSalt;
}
